flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/SortUtil.scala

```diff
@@ -0,0 +1,353 @@
+import org.apache.calcite.rel.`type`._
+import org.apache.calcite.rel.logical.LogicalSort
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.calcite.sql.`type`.SqlTypeName
+import org.apache.flink.table.api.TableException
+import org.apache.calcite.sql.`type`.SqlTypeName
+import org.apache.calcite.sql.`type`.SqlTypeName._
+import org.apache.flink.table.functions.Accumulator
+import java.util.{ List => JList, ArrayList }
+import org.apache.flink.api.common.typeinfo.{ SqlTimeTypeInfo, TypeInformation }
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import java.sql.Timestamp
+import org.apache.calcite.rel.RelFieldCollation
+import org.apache.calcite.rel.RelFieldCollation.Direction
+import java.util.Comparator
+import org.apache.flink.api.common.typeutils.TypeComparator
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo._
+import java.lang.{Byte=>JByte,Integer=>JInt,Long=>JLong,Double=>JDouble,Short=>JShort,String=>JString,Float=>JFloat}
+import java.math.{BigDecimal=>JBigDecimal}
+import org.apache.flink.api.common.functions.MapFunction
+
+/**
+ * Class represents a collection of helper methods to build the sort logic.
+ * It encapsulates as well the implementation for ordering and generic interfaces
+ */
+
+object SortUtil {
+
+  /**
+   * Function creates [org.apache.flink.streaming.api.functions.ProcessFunction] for sorting 
+   * elements based on proctime and potentially other fields
+   * @param calcSort Sort logical object
+   * @param inputType input row type
+   * @return org.apache.flink.streaming.api.functions.ProcessFunction
+   */
+  private[flink] def createProcTimeSortFunction(
+    calcSort: LogicalSort,
+    inputType: RelDataType): ProcessFunction[Row, Row] = {
+
+    val keySortFields = getSortFieldIndexList(calcSort)
+    val keySortDirections = getSortFieldDirectionList(calcSort)
+
+    val inputRowType = FlinkTypeFactory.toInternalRowTypeInfo(inputType).asInstanceOf[RowTypeInfo]
+    
+    val orderings = createOrderingComparison(inputType, keySortFields, keySortDirections)
+    
+    //drop time from comparison
+    val keyIndexesNoTime = keySortFields.slice(1, keySortFields.size)
+    val orderingsNoTime = orderings.slice(1, keySortFields.size)
+    
+    val rowComparator = createRowSortComparator(keyIndexesNoTime,orderingsNoTime)
+    
+    new ProcTimeBoundedSortProcessFunction(
+      inputType.getFieldCount,
+      inputRowType,
+      rowComparator)
+
+  }
+
+   /**
+   * Function creates a row comparator for the sorting fields based on
+   * [java.util.Comparator] objects derived from [org.apache.flink.api.common.TypeInfo]
+   * @param keyIndex the indexes of the fields on which the sorting is done. 
+   * First is expected to be the time  
+   * @param orderings the [UntypedOrdering] objects 
+   * @return Array of ordering objects
+   */
+  def createRowSortComparator(keyIndex: Array[Int],
```

Comment:
@fhueske  It seems that RowComparator is not extending the java.util.Comparator ...so we need to wrap it in order to use it directly with the sorting algorithms from the collection library